### ISSUE TITLE
Strange error message && more segfaults.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2607]
Migrated from [rt.perl.org#4281](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=4281) (status was 'resolved')

Searchable as RT4281$

**Comment by p5pRT** [comment]
### From @abigail

#### Created by @abigail



&nbsp;    \#\!/opt/perl/bin/perl \-w

&nbsp;    sub foo \(\) \{
&nbsp;        "" =~ qr /\(?\{ return \}\)/;
&nbsp;        return "End\\n";
&nbsp;    \}

&nbsp;    print foo \(\);
&nbsp;    \_\_END\_\_
&nbsp;    Can't return outside a subroutine at y\.pl line 5\.


Removing the return leads to a segmentation fault&#8203;:


&nbsp;    \#\!/opt/perl/bin/perl \-w

&nbsp;    sub foo \(\) \{
&nbsp;        "" =~ qr /\(?\{ return \}\)/;
&nbsp;    \}

&nbsp;    print foo \(\);
&nbsp;    \_\_END\_\_
&nbsp;    Segmentation fault
&nbsp;    

Identical results with 5\.7\.0\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by abigail at Wed Jun 14 21:00:02 EDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.13, archname=i686-linux-64int
    uname='linux alexandra 2.2.13 #5 tue feb 8 15:37:54 est 2000 i686 unknown '
    config_args='-Dprefix=/opt/perl -d -Uinstallusrbinperl -Doptimize=-g -Dusemorebits'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=define use64bitall=undef uselongdouble=define usesocks=undef
  Compiler:
    cc='cc', optimize='-g', gccversion=2.95.2 19991024 (release)
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /home/abigail/Perl
    /home/abigail/Sybase
    /opt/perl/lib/5.6.0/i686-linux-64int
    /opt/perl/lib/5.6.0
    /opt/perl/lib/site_perl/5.6.0/i686-linux-64int
    /opt/perl/lib/site_perl/5.6.0
    /opt/perl/lib/site_perl/5.005
    /opt/perl/lib/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/abigail
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/home/abigail/Lib:/usr/local/lib:/usr/lib:/lib:/usr/X11R6/lib:/opt/tcl/lib:/opt/tk/lib/tk8.0
    LOGDIR (unset)
    PATH=/home/abigail/Bin:/opt/perl/bin:/opt/tcl/bin:/opt/tk/bin:/usr/local/bin:/usr/local/X11/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/X11R6/bin:/usr/games:/opt/povray/bin:/opt/teTeX/bin/i686-pc-linux-gnu:/opt/python/bin
    PERL5LIB=/home/abigail/Perl:/home/abigail/Sybase
    PERLDIR=/opt/perl
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

abigail writes&#8203;:
>     sub foo \(\) \{
>         "" =~ qr /\(?\{ return \}\)/;
>         return "End\\n";
>     \}
> 
>     print foo \(\);
>     \_\_END\_\_
>     Can't return outside a subroutine at y\.pl line 5\.
> 
> 
> Removing the return leads to a segmentation fault&#8203;:
> 
> 
>     \#\!/opt/perl/bin/perl \-w
> 
>     sub foo \(\) \{
>         "" =~ qr /\(?\{ return \}\)/;
>     \}
> 
>     print foo \(\);
>     \_\_END\_\_
>     Segmentation fault

I'd suspect that the \<"" =~ qr /\(?\{ return \}\)/;> line has hosed up the sub
stack so that perl doesn't think it is in a sub anymore\.  Witness that the
code&#8203;:

&nbsp;    sub foo \(\) \{
&nbsp;        "" =~ qr /\(abc\)/;
&nbsp;        return "End\\n";
&nbsp;    \}
&nbsp;    
&nbsp;    print foo \(\);

will print "End" just fine\. I'd look at the \<qr /\(?\{ return \}\)/> myself\, as
the "\(?\{ return \}\)" construct seems to be what is hosing things up\.

Purify trace is after my \.sig\.
====================================================
Mark Leighton Fisher    Thomson Consumer Electronics
fisherm@&#8203;tce\.com         Indianapolis\, IN\, USA
"Display some adaptability\."  \-\- Doug Shaftoe\, \_Cryptonomicon\_

\[E\] IPR&#8203;: Invalid pointer read in Perl\_pp\_leavesub \{1 occurrence\}
&nbsp;        Reading 1 byte from 0x18238732 \(1 byte at 0x18238732 illegal\)
&nbsp;        Address 0x18238732 points into a HeapAlloc'd block in unallocated
region of heap 0x18230000
&nbsp;        Thread ID&#8203;: 0x192
&nbsp;        Error location
&nbsp;            Perl\_pp\_leavesub \[pp\_hot\.c&#8203;:2044\]
&nbsp;                    \}
&nbsp;                    PUTBACK;
&nbsp;                
&nbsp;             =>     POPSUB\(cx\,sv\);    /\* Stack values are safe&#8203;: release CV
and @&#8203;\_ \.\.\. \*/
&nbsp;                    PL\_curpm = newpm;    /\* \.\.\. and pop $1 et al \*/
&nbsp;                
&nbsp;                    LEAVE;
&nbsp;            Perl\_runops\_debug \[run\.c&#8203;:56\]
&nbsp;                        DEBUG\_t\(debop\(PL\_op\)\);
&nbsp;                        DEBUG\_P\(debprof\(PL\_op\)\);
&nbsp;                    \}
&nbsp;             =>     \} while \(\(PL\_op = CALL\_FPTR\(PL\_op\->op\_ppaddr\)\(aTHX\)\)\);
&nbsp;                
&nbsp;                    TAINT\_NOT;
&nbsp;                    return 0;
&nbsp;            S\_run\_body     \[perl\.c&#8203;:1401\]
&nbsp;                    else if \(PL\_main\_start\) \{
&nbsp;                    CvDEPTH\(PL\_main\_cv\) = 1;
&nbsp;                    PL\_op = PL\_main\_start;
&nbsp;             =>     CALLRUNOPS\(aTHX\);
&nbsp;                    \}
&nbsp;                
&nbsp;                    my\_exit\(0\);
&nbsp;            perl\_run       \[perl\.c&#8203;:1325\]
&nbsp;            RunPerl        \[perllib\.c&#8203;:319\]
&nbsp;            main           \[perlmain\.c&#8203;:18\]
&nbsp;            mainCRTStartup \[crtexe\.obj\]

**Comment by p5pRT** [comment]
### From @vanstyn

In \<20000911072001\.21898\.qmail@&#8203;foad\.org>\, abigail@&#8203;foad\.org writes&#8203;:
:This is a bug report for perl from abigail@&#8203;foad\.org\,
:generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
:
:
:\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
:\[Please enter your report here\]
:
:    \#\!/opt/perl/bin/perl \-w
:
:    sub foo \(\) \{
:        "" =~ qr /\(?\{ return \}\)/;
:        return "End\\n";
:    \}
:
:    print foo \(\);
:    \_\_END\_\_
:    Can't return outside a subroutine at y\.pl line 5\.
:
:
:Removing the return leads to a segmentation fault&#8203;:

I believe this is a symptom of the known problems fitting the
concept of re\_eval into the call stack\. I'm not aware that anyone
has a solution to this\, but I think one of the conclusions was
that the whole JMPENV approach was fundamentally flawed\.

Hugo

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'stalled' to 'open'

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

On Sun Sep 10 17&#8203;:22&#8203;:33 2000\, abigail wrote&#8203;:
> This is a bug report for perl from abigail@&#8203;foad\.org\,
> generated with the help of perlbug 1\.28 running under perl v5\.6\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
>     \#\!/opt/perl/bin/perl \-w
> 
>     sub foo \(\) \{
>         "" =~ qr /\(?\{ return \}\)/;
>         return "End\\n";
>     \}
> 
>     print foo \(\);
>     \_\_END\_\_
>     Can't return outside a subroutine at y\.pl line 5\.

$ cat rt\-4281\_a\.pl
\#\!/opt/perl/bin/perl \-wl

sub foo \(\) \{
"" =~ qr /\(?\{ return "a" \}\)/;
return "End\\n";
\}

print foo \(\);
\_\_END\_\_


perl\-5\.6\.2 rt\-4281\_a\.pl 
a
Can't return outside a subroutine at rt\-4281\_a\.pl line 5\.

perl\-5\.8\.0 rt\-4281\_a\.pl 
a
Can't return outside a subroutine at rt\-4281\_a\.pl line 5\.

perl\-5\.8\.8 rt\-4281\_a\.pl 
a
Can't return outside a subroutine at rt\-4281\_a\.pl line 5\.

perl\-5\.10\.0 rt\-4281\_a\.pl 
a
Can't return outside a subroutine at rt\-4281\_a\.pl line 5\.

perl\-blead rt\-4281\_a\.pl 
a
Can't return outside a subroutine at rt\-4281\_a\.pl line 5\.



> Removing the return leads to a segmentation fault&#8203;:
> 
> 
>     \#\!/opt/perl/bin/perl \-w
> 
>     sub foo \(\) \{
>         "" =~ qr /\(?\{ return \}\)/;
>     \}
> 
>     print foo \(\);
>     \_\_END\_\_
>     Segmentation fault
> 

$ cat rt\-4281\_b\.pl
\#\!/opt/perl/bin/perl \-wl

sub foo \(\) \{
"" =~ qr /\(?\{ return "b" \}\)/;
\}

print foo \(\);
\_\_END\_\_

perl\-5\.6\.2 rt\-4281\_b\.pl 
b
Segmentation fault

perl\-5\.8\.0 rt\-4281\_b\.pl 
b
Segmentation fault

perl\-5\.8\.1 rt\-4281\_b\.pl 
b

perl\-5\.10\.0 rt\-4281\_b\.pl 
b

perl\-blead rt\-4281\_b\.pl 
b


So the segfault is gone \(or hiding\) but the incorrect\(?\) warning is 
still there\.

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
